Rich-text support for comments and response text #5830

belyabl9 opened the pull request:
Fixes #5830

belyabl9 commented:
Ready to review

wkurniawan07 commented:
@belyabl9 correct me if I'm wrong, but this PR includes support for response texts as well?

belyabl9 commented:
@wkurniawan07 , you're right. I'll change the issue and PR name

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorCommentsPageUiTest.java:
```diff
@@ -179,6 +180,7 @@ private void testPanelsCollapseExpand() {
     private void testActions() throws Exception {
         ______TS("action: edit student comment");
         commentsPage.clickStudentCommentEditForRow(1);
+        ThreadHelper.waitFor(1000);
```
Fixed time waits like this are not preferred in general. What is this line waiting for?

belyabl9 replied:
It waits for TinyMCE initialization, because lazy initialization is used i.e. only after showing edit row

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java:
```diff
@@ -43,7 +43,8 @@ public void fillResponseTextBox(int qnNumber, int responseNumber, String text) {
         click(element);
-        fillTextBox(element, text);
+        element.clear();
+        element.sendKeys(text);
```
Why does the previous `fillTextBox(...)` not work?

belyabl9 replied:
It should be enough. Initially, I changed it to fillRichTextEditor() but such approach doesn't handle symbols counter, so I came back. I'll use fillTextBox(), it seems to be the same

wkurniawan07 commented:
Two things:
- Using `ThreadHelper.waitFor(...)` is in general discouraged. The preferred way is to add explicit wait for the expected condition. There are many examples of this in `AppPage.java`.
- If there are `element.click()`, let's change it to `click(element)` so that Selenium can fully support it, unless there are good reasons to retain the first one?

belyabl9 commented:
Ready to review

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/shared/commentRow.tag:
```diff
@@ -262,12 +262,12 @@
-                              id="commentText${commentIndex}">${comment.commentText}</textarea>
+                    <div id="rich-text-toolbar-comment-container-${commentIndex}"></div>
+                    <div class="panel panel-default panel-body"
+                         id="commentText${commentIndex}">
```
This should be fine on the previous line.

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag:
```diff
@@ -189,11 +189,13 @@
+        <div id="rich-text-toolbar-comment-container-${divId}"></div>
+        <div class="panel panel-default panel-body"
+             rows="3"
+             id="${textAreaId}-${divId}">
```
Here as well, in fact starting from `rows="3"`.

wkurniawan07 commented on src/main/webapp/js/common.js:
```diff
@@ -884,6 +884,16 @@ function hideSingleCollapse(e) {
     $(heading).find('a.btn').hide();
 }
 
+function destroyEditor(id) {
```
While we're at it, let's add a header comment for this function.

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java:
```diff
@@ -174,11 +175,11 @@ public void fillTextareaToEditResponseComment(int sessionIdx, int questionIdx, i
-        textarea.sendKeys(text);
+                By.id(editorId));
+        waitForRichTextEditorToLoad(editorId);
+        fillTextBox(textarea, text);
```
Why is this not `fillRichTextEditor`?

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java:
```diff
@@ -174,11 +175,11 @@ public void fillTextareaToEditResponseComment(int sessionIdx, int questionIdx, i
-        click(textarea);
-        textarea.clear();
-        textarea.sendKeys(text);
+                By.id(editorId));
```
Again this one should be fine on the previous line.

wkurniawan07 commented on src/test/resources/pages/instructorCommentsPageAddFrc.html:
```diff
@@ -1809,7 +1839,7 @@ <h4 class="col-sm-9 text-color-primary">
                           </div>
                           <div id="plainCommentText-1-1-1-1" style="margin-left: 15px;">
-                            added response comment
+                            Instructor 1 comment to student 1 self feedback
```
Any idea why the previous response comment is removed?

wkurniawan07 commented on src/test/resources/pages/instructorCommentsPageDeleteFrc.html:
```diff
@@ -1744,9 +1778,9 @@ <h4 class="col-sm-9 text-color-primary">
               <div class="panel-heading">
                 <b>
-                  Question 1
+                  Question 2
```
The first question seems to get removed. I suspect it's an off-by-one error somewhere?

wkurniawan07 commented:
Added some comments, the most important one being there are functions that might have been broken. Also there are merge conflicts to be resolved.

belyabl9 commented:
@wkurniawan07 , I'll check it in more detail, but I think that the cause of removed questions is in InstructorStudentCommentEditAction.java. I removed assertion for deletion with empty comment text, can't get why it was needed if comment has identifier

wkurniawan07 commented on src/test/resources/pages/instructorSearchPageSearchStudentComments.html:
```diff
@@ -135,7 +135,7 @@
                   [Tue, 01 May 2012, 08:49 AM UTC]
                 </span>
-                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="display:none;" target="_blank" title="" type="button">
+                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="" target="_blank" title="" type="button">
```
This change should not occur. I have tried this branch on my own local environment.

wkurniawan07 commented on src/test/resources/pages/instructorSearchPageSearchComments.html:
```diff
@@ -135,7 +135,7 @@
                   [Tue, 01 May 2012, 08:49 AM UTC]
                 </span>
-                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="display:none;" target="_blank" title="" type="button">
+                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="" target="_blank" title="" type="button">
```
Similarly here.

belyabl9 replied:
I have a problem with it. Actual result is variable

wkurniawan07 replied:
I can't replicate the problem, it's consistently returning the correct one (with `display:none;`) for me. Same with your other PR.

belyabl9 replied:
It's strange. What about FF version, as I remember it was updated for testing. I am still on previous one. May it be the reason ?

wkurniawan07 replied:
Unlikely, because I just tried with the old 38.0.5 and it still consistently returns the correct one.
Anyway, let's move forward with this PR: undo these changes, they're preventing a passing build in Travis.

wkurniawan07 commented on src/test/resources/pages/instructorCommentsPageAddSc.html:
```diff
@@ -227,7 +227,7 @@ <h4 class="col-sm-9 text-color-primary">
                         all students in this course
                       </b>
-                      [Fri, 01 Jun 2012, 07:20 PM UTC] (last edited by comments.instructor1@course1.tmt at ${datetime.now})
+                      [Fri, 01 Jun 2012, 07:20 PM UTC] (last edited by comments.instructor1@course1.tmt at Thu, 28 Jul 2016, 09:54 PM UTC)
```
You need to run the tests with `-Duser.timezone=UTC` VM args, otherwise there are certain times in the day where this kind of unwanted replacement happens.

belyabl9 commented:
Ready to review

wkurniawan07 commented:
@belyabl9 I've identified what caused the missing comment in the `instructorCommentsPageEditFrc.html`. It doesn't look like anything to do with Firefox cache or alike. As it happens, this action on line 230 of `InstructorCommentsPageUiTest`:
```java
commentsPage.saveResponseComment(1, 1, 1, 1);
```
actually did not execute and returns "Comment cannot be empty" for some reason. Can you fix that?

belyabl9 commented:
@wkurniawan07 , I've fixed that issue. Ready to review

belyabl9 commented:
Please, review

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/CommentAttributes.java:
```diff
@@ -48,7 +48,7 @@ public CommentAttributes(String courseId, String giverEmail, CommentParticipantT
         this.recipientType = recipientType == null ? CommentParticipantType.PERSON : recipientType;
         this.recipients = recipients;
-        this.commentText = commentText;
+        this.commentText = commentText == null ? null : new Text(Sanitizer.sanitizeForRichText(commentText.getValue()));
```
Just thinking, is it possible to make a `sanitizeForRichText` method that takes in `Text` parameter? It can be reused here and potentially many other places.

belyabl9 replied:
As it's used in other places unrelated to this issue, it'd make sense to do it in a separate issue

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/shared/commentRow.tag:
```diff
@@ -262,12 +262,11 @@
+                    <div class="panel panel-default panel-body" id="commentText${commentIndex}">
+                         ${comment.commentText}
+                    </div>
+                    <input type="hidden" name="commenttext">
```
`commenttext` should be taken from a constant in `Const.ParamsNames`, right?

wkurniawan07 commented on src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag:
```diff
@@ -189,11 +189,11 @@
+        <div class="panel panel-default panel-body" id="${textAreaId}-${divId}">
+            ${frc.commentText}
+        </div>
+        <input type="hidden" name="responsecommenttext">
```
Similarly for `responsecommenttext`.

wkurniawan07 commented on src/main/webapp/js/common.js:
```diff
@@ -887,6 +887,19 @@ function hideSingleCollapse(e) {
 /**
+ * Destroys an instance of TinyMCE rich-text editor.
+ */
+function destroyEditor(id) {
```
Move this to `richTextEditor.js`?

wkurniawan07 commented on src/main/webapp/js/omniComment.js:
```diff
@@ -402,6 +410,16 @@ function enableComment(commentIdx) {
+            fixed_toolbar_container: '#rich-text-toolbar-comment-container-' + commentIdx
+        });
+        /* eslint-enable camelcase */
+    }
```
Actually this routine looks repeated many times as well. Can they be extracted to `richTextEditor.js`?

belyabl9 replied:
I think it can't. It's already a routine in richTextEditor.js and it has variable params like 'fixed_toolbar_container' and selector

wkurniawan07 commented:
Added a few comments to address, some possibly not in this PR.

belyabl9 commented:
Ready to review

belyabl9 commented:
Please, review

wkurniawan07 commented on src/main/webapp/js/contextualcomments.js:
```diff
@@ -4,16 +4,25 @@ $(document).ready(function() {
+    }
+
     $('form[name="form_commentadd"]').submit(function() {
+        tinymce.get('commenttext').save();
```
What's the difference between `tinymce` and `tinyMCE`?

belyabl9 replied:
As I understand, both variants are possible, like aliases

wkurniawan07 commented on src/test/resources/pages/InstructorCourseDetailsEmptyCourse.html:
```diff
@@ -1,5 +1,8 @@
 <head>
+<style id="mceDefaultStyles" type="text/css">
+  .mce-content-body div.mce-resizehandle {position: absolute;border: 1px solid black;background: #FFF;width: 5px;height: 5px;z-index: 10000}.mce-content-body .mce-resizehandle:hover {background: #000}.mce-content-body img[data-mce-selected], hr[data-mce-selected] {outline: 1px solid black;resize: none}.mce-content-body .mce-clonedresizable {position: absolute;opacity: .5;filter: alpha(opacity=50) ...
+</style>
```
You need to regenerate this HTML file again, a placeholder `${tinymce.style}` is expected here due to #5934. Similarly for other files that have these 3 lines.

wkurniawan07 replied:
These lines will be ignored anyway, so can you remove them? Similarly for the other files.

wkurniawan07 commented:
Changes are good, just that one minor thing needs to be done on some HTML test files.

belyabl9 commented:
@wkurniawan07 , please, review
Travis results are a little bit strange

wkurniawan07 commented:
@belyabl9 apologies, there's little I can do here until Travis decides to cooperate with us again.

wkurniawan07 commented:
@belyabl9 can you sync this branch with the current `master`? I need to see if the problem is still there.

belyabl9 commented:
@wkurniawan07 , done. As I see, the issue persists

damithc commented:
Can this be wrapped up soon?

belyabl9 commented:
@wkurniawan07 promised (in slack) to have a look at this strange Travis error, so I'll proceed after his comment

wkurniawan07 commented:
I'm at a loss for the behaviour myself. Since we don't really have the luxury of time, I'll be circumventing this problem by ignoring the guilty placeholder. Putting this on hold for a very short while, will update later.

wkurniawan07 commented:
@belyabl9 this can be continued now

belyabl9 commented:
Ready to review

wkurniawan07 commented:
Everything looks good except for that possible off-by-one error (dated more than a month back). If it's indeed caused by the removed assertion, can you revert it for now? It's better not to have unexpected behavior changes at this point of time. If it's not the cause, then it's something that needs to be fixed.

damithc commented:
@belyabl9 for your followup please.

belyabl9 commented:
I'll try to check it this week. My current schedule is more tight than in summer

wkurniawan07 commented:
@belyabl9 any updates here?

belyabl9 commented:
@wkurniawan07 , I'll continue to reproduce this issue today, By the way, as I see from the comments history I've already fixed that.

wkurniawan07 commented:
@belyabl9 it has not been fixed yet. If you are referring to [this comment](#issuecomment-237155611), it refers for some failure in `instructorCommentsPageEditFrc.html` which indeed has been fixed, but the issues for `instructorCommentsPageAddFrc.html` and `instructorCommentsPageDeleteFrc.html` still persist.

belyabl9 commented:
@wkurniawan07 , I'll have a look on it today. As I remember, I checked it previously and it appeared only via running test suite. In a debug mode it worked as expected

belyabl9 commented:
@wkurniawan07 , as I previously told, when I ran all the tests, I got the previous result. When I run single test (InstructorCommentsPageUiTest), I get the diff as in the last commit. The most interesting thing for me is that Travis passed both variants.
So, please, have a look at the last commit

belyabl9 commented:
Ready to review

wkurniawan07 commented:
@belyabl9 I've somewhat isolated the cause of the strange problem: it seems like the test driver cannot access the rich text editor in the comments page if the browser is not active (e.g in background). Can you try to fix it?

wkurniawan07 commented:
Hi @belyabl9, any updates on this PR?

belyabl9 commented:
Guys, sorry for delay. My ISP stopped providing its services, so currently I'm looking for a new one.
I'll return as soon as I restore stable internet connection

wkurniawan07 commented:
Hi @belyabl9, any updates on this PR and/or your ISP status?

belyabl9 commented:
@wkurniawan07 , new provider promised to connect me this week, I'll notify ASAP

wkurniawan07 commented:
@belyabl9 thanks for the update. While waiting, can you try your best to work on this offline? TEAMMATES has been configured to be able to be developed fully offline, save for one UI test which I don't think is affected by this PR.

belyabl9 commented:
@wkurniawan07 , I'm an internet user again :) I've rebased the code, fixed some found bugs.
Could you, please, review it and point me out on the issues which still hold this code from being merged

Ready to review

wkurniawan07 commented:
@belyabl9 sadly the problem that I pointed out before still persists: if `InstructorCommentsPageUiTest` is running while the Firefox browser is on the background (meaning you work on something else on the foreground when the test is running), it will fail. Seems like the Selenium driver cannot access the rich text editor, although it might also be a javascript issue. Can you try your best to fix it?

belyabl9 commented:
@wkurniawan07 , I can't reproduce the described issue. Here are my steps:
1. Run InstructorCommentsPageUiTest (firefox)
2. Surf the web via Chrome on the foreground
3. Check the result after some time
Is it the same you are doing to reproduce it ? Could you add more details, because I have successful result regardless of firefox is in the foreground or background

damithc commented:
Perhaps the issue is OS specific?

belyabl9 commented:
ping

wkurniawan07 commented:
@belyabl9 sorry for the delay, because this is one of the PRs that will take some time to review. I will now be continuing with the review.
I shall try running on my Windows machine. On my Mac and Travis CI which is running Linux the failure is confirmed, and I'd very much prefer if it can be fixed because it may be affecting the CI run in the long run.

Also, there are merge conflicts that needs resolve.

wkurniawan07 commented on src/test/resources/pages/allJsUnitTests.html:
```diff
@@ -8,7 +8,9 @@
-  <script type="text/javascript" src="../../../main/webapp/js/common.js" data-cover></script>  
+  <script type="text/javascript" src="../../../main/webapp/js/richTextEditor.js" data-cover></script>
+  <script type="text/javascript" src="../../../main/webapp/js/common.js" data-cover></script>
+  <script type="text/javascript" src="../../../main/webapp/js/lib/tinymce.min.js"></script>
```
This should come before `richTextEditor.js`.

wkurniawan07 commented:
It's reproducible in my Windows machine so it doesn't seem to be OS-specific, but we may take a look on it on another issue.

belyabl9 commented:
@wkurniawan07 , ok. I'll do it on Friday
